internal/lsp/bundles/cache.go

```diff
@@ -0,0 +1,238 @@
+package bundles
+
+import (
+	"bytes"
+	//nolint:gosec
+	"crypto/md5"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/open-policy-agent/opa/bundle"
+)
+
+// Cache is a struct that maintains a number of bundles in memory and
+// provides a way to refresh them when the source files change.
+type Cache struct {
+	workspacePath string
+	bundles       map[string]*cacheBundle
+	errorLog      io.Writer
+}
+
+type CacheOptions struct {
+	WorkspacePath string
+	ErrorLog      io.Writer
+}
+
+func NewCache(opts *CacheOptions) *Cache {
+	workspacePath := opts.WorkspacePath
+
+	if !strings.HasSuffix(workspacePath, string(filepath.Separator)) {
+		workspacePath += string(filepath.Separator)
+	}
+
+	c := &Cache{
+		workspacePath: workspacePath,
+		bundles:       make(map[string]*cacheBundle),
+	}
+
+	if opts.ErrorLog != nil {
+		c.errorLog = opts.ErrorLog
+	}
+
+	return c
+}
+
+// Refresh walks the workspace path and loads or refreshes any bundles that
+// have changed since the last refresh.
+func (c *Cache) Refresh() ([]string, error) {
+	if c.workspacePath == "" {
+		return nil, errors.New("workspace path is empty")
+	}
+
+	// find all the bundle roots that are currently present on disk
+	var foundBundleRoots []string
+
+	err := filepath.Walk(c.workspacePath, func(path string, info os.FileInfo, _ error) error {
+		if info.IsDir() && (info.Name() == ".git" || info.Name() == ".idea") {
+			return filepath.SkipDir
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		if filepath.Base(path) == ".manifest" {
+			foundBundleRoots = append(
+				foundBundleRoots,
+				strings.TrimPrefix(filepath.Dir(path), c.workspacePath),
+			)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to walk workspace path: %w", err)
+	}
+
+	var refreshedBundles []string
+
+	// refresh any bundles that have changed
+	for _, root := range foundBundleRoots {
+		if _, ok := c.bundles[root]; !ok {
+			c.bundles[root] = &cacheBundle{}
+		}
+
+		refreshed, err := c.bundles[root].Refresh(filepath.Join(c.workspacePath, root))
+		if err != nil {
+			if c.errorLog != nil {
+				fmt.Fprintf(c.errorLog, "failed to refresh bundle %q: %v\n", root, err)
+			}
+
+			continue
+		}
+
+		if refreshed {
+			refreshedBundles = append(refreshedBundles, root)
+		}
+	}
+
+	// remove any bundles that are no longer present on disk
+	for root := range c.bundles {
+		found := false
+
+		for _, foundRoot := range foundBundleRoots {
+			if root == foundRoot {
+				found = true
+
+				break
+			}
+		}
+
+		if !found {
+			delete(c.bundles, root)
+		}
+	}
+
+	return refreshedBundles, nil
+}
+
+// List returns a list of all the bundle roots that are currently present in
+// the cache.
+func (c *Cache) List() []string {
+	roots := make([]string, 0, len(c.bundles))
+	for root := range c.bundles {
+		roots = append(roots, root)
+	}
```

Comment:
Could use `util.Keys` :)

Reply:
Very nice, done in 9d2aa3bfad3159476acf2b8ba9ddf0e47cddde9a